### PR TITLE
Client library updates for Dunning Campaign API enhancements

### PR DESCRIPTION
**Pull request by amandamfielding**
* Add `dunning_campaign_id` to `Account`, `Invoice`, and `Plan` 
* PUT v2/dunning_campaigns/{dunning_campaign_id}/bulk_update
```
campaign = Recurly::DunningCampaign.find('id')
campaign.bulk_update(plan_codes)
```

### CODE REVIEW

**Comment by sunnyw1212** [comment]
just a reminder that we'll need to wait for https://github.com/recurly/recurly-app/pull/17994 to be merged and implement

```
- Changed Account object response:
- Added dunning_campaign_id
- Changed POST v2/accounts and PUT v2/accounts/{account_code}
- Added <dunning_campaign_id> field
- Changed POST v2/purchases, POST /v2/purchases/preview and POST v2/purchases/authorize requests
- Added <dunning_campaign_id> field
- Changed POST v2/subscriptions and POST v2/subscriptions/preview requests
- Added <dunning_campaign_id> field
```

before we merge this cahnge

File: lib/recurly/dunning_campaign.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+      description
+      default
+      dunning_cycles
+      created_at
+      updated_at
+      deleted_at
+      plans
+    )
+
+    def bulk_update(plan_codes)
```
Comment (sunnyw1212) [inline lib/recurly/dunning_campaign.rb:19]: do we also need to support `plan_ids` here?

File: lib/recurly/dunning_campaign.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      return false unless link? :bulk_update
```
Comment (sunnyw1212) [inline lib/recurly/dunning_campaign.rb:20]: can you explain what this is doing?

File: lib/recurly/dunning_campaign.rb [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+      builder = XML.new("<dunning_campaign/>")
+      node = builder.add_element("plan_codes")
+      plan_codes.each { |plan_code| node.add_element "plan_code", plan_code }
+
+      reload follow_link(:bulk_update, :body => builder.to_s)
```
Comment (sunnyw1212) [inline lib/recurly/dunning_campaign.rb:26]: can you explain what this is doing?

File: lib/recurly/resource.rb [hunk 4]
```diff
@@ -608,7 +608,7 @@ def embedded!(root_index = false)
 
       def find_resource_class(name)
         resource_name = Helper.classify(name)
-        if Recurly.const_defined?(resource_name, false)
+        if Recurly.const_defined?(resource_name, false) && Recurly.const_get(resource_name, false).instance_of?(Class)
```
Comment (sunnyw1212) [inline lib/recurly/resource.rb:611]: can you explain what this is doing?

Comment (judith) [reply lib/recurly/resource.rb:611]: There is both an attribute `version` on the `DunningCycle` resource and a class `Version` this additional check makes sure that when it is looking for the resource class that it is also verifying that the constant is for the class and not an attribute with the same name.

File: lib/recurly/dunning_campaign.rb (hunk 1 above)
Comment (judith) [reply lib/recurly/dunning_campaign.rb:19]: For API V2 the `plan_code` is the attribute for looking up a `Plan` resource. The `id` field is not exposed for plans in API V2.

File: lib/recurly/dunning_campaign.rb (hunk 2 above)
Comment (judith) [reply lib/recurly/dunning_campaign.rb:20]: It is checking if there is an action link for `:bulk_update` as part of the response when instantiating the resource.

File: lib/recurly/dunning_campaign.rb (hunk 3 above)
Comment (judith) [reply lib/recurly/dunning_campaign.rb:26]: It is making a request to the `href` with the `method` in the action link for that resource with the xml body for sending the plans and then creating a new instance of this resource from the xml response for that API request.

File: lib/recurly/invoice.rb [hunk 5]
```diff
@@ -115,6 +115,7 @@ def invoice_number_with_prefix
       surcharge_in_cents
       tax_details
       billing_info_uuid
+      dunning_campaign_id
```
Comment (sunnyw1212) [inline lib/recurly/invoice.rb:118]: i think we only want dunning_campaign_id to be readable and NOT writable - do we need to indicate that differently somehow here?

Comment (judith) [reply lib/recurly/invoice.rb:118]: That should be handled in server side validation for the API request. No indication is needed in the client library.

**Review by judith — APPROVED** [review]

File: lib/recurly/dunning_cycle.rb [hunk 6]
```diff
@@ -0,0 +1,18 @@
+module Recurly
+  class DunningCycle < Resource
+    define_attribute_methods %w(
+      type
+      applies_to_manual
```
Comment (judith) [inline lib/recurly/dunning_cycle.rb]: This should be `applies_to_manual_trial`.

**Review by judith — APPROVED** [review]

**Pull request merged**